Q1.java

```diff
@@ -0,0 +1,95 @@
+import java.util.*;
+
+class Q1
+{
+	public static void main(String[] args)
+	{
+		int[] start = new int[4];
+		int[] end = new int[4];
+
+		start[0] = 1;
+		start[1] = 2;
+		start[2] = 0;
+		start[3] = 3;
+
+		end[0] = 3;
+		end[1] = 1;
+		end[2] = 2;
+		end[3] = 0;
+
+		rearrange(start, end);
+	}
+
+
+	public static void rearrange(int[] start, int[] end)
+	{
+		HashMap<Integer, Integer> positions = calculateMap(start);
+
+		for(int i=0; i<start.length; i++)
+		{
+
+			if(start[i] != end[i] || end[i] != 0)
+			{
+				move0(start, positions, i);
+
+				int wantedCar = end[i];
+				int positionOfCar = positions.get(wantedCar);
+
+				//swapping places in array
+				start[i] = wantedCar;
+				start[positionOfCar] = 0;
+
+				//altering the map
+				positions.put(wantedCar, i);
+				positions.put(0, positionOfCar);
+
+				System.out.println("swap positions " + i + " and " + positionOfCar);
```

Comment:
You can make this code reusable by defining class `Move` and make your `rearrange` method return `List<Move>`. This will this code also testable.